Unicode in messages throws internal server error

Writing unicode as as a message throws an internal server error. SecureDrop should support unicode in messages for those of us whose instances are used by non-English speakers. Not only does this happen, we get the generic "internal server error" message that doesn't explain what went wrong or how. This is very bad for sources who upon reaching a SD instance will be unable to submit, not see why, and presumably assume SD isn't a service worth using.

Moving to Python 3 (#997) is probably the easiest way to do this since we get better unicode handling for minimal overhead.

I haven't had time to install the German TBB, but this is the traceback I'm getting with my attempted Python 3 implementation. Hopefully I can get to that this weekend.

```
tests/test_unit_source.py:246: in test_submit_sanitizes_filename
    ), follow_redirects=True)
/usr/local/lib/python3.4/dist-packages/werkzeug/test.py:788: in post
    return self.open(*args, **kw)
/usr/local/lib/python3.4/dist-packages/flask/testing.py:113: in open
    follow_redirects=follow_redirects)
/usr/local/lib/python3.4/dist-packages/werkzeug/test.py:751: in open
    response = self.run_wsgi_app(environ, buffered=buffered)
/usr/local/lib/python3.4/dist-packages/werkzeug/test.py:668: in run_wsgi_app
    rv = run_wsgi_app(self.application, environ, buffered=buffered)
/usr/local/lib/python3.4/dist-packages/werkzeug/test.py:871: in run_wsgi_app
    app_rv = app(environ, start_response)
/usr/local/lib/python3.4/dist-packages/flask/app.py:2000: in __call__
    return self.wsgi_app(environ, start_response)
/usr/local/lib/python3.4/dist-packages/flask/app.py:1991: in wsgi_app
    response = self.make_response(self.handle_exception(e))
/usr/local/lib/python3.4/dist-packages/flask/app.py:1567: in handle_exception
    reraise(exc_type, exc_value, tb)
/usr/local/lib/python3.4/dist-packages/flask/_compat.py:33: in reraise
    raise value
/usr/local/lib/python3.4/dist-packages/flask/app.py:1988: in wsgi_app
    response = self.full_dispatch_request()
/usr/local/lib/python3.4/dist-packages/flask/app.py:1641: in full_dispatch_request
    rv = self.handle_user_exception(e)
/usr/local/lib/python3.4/dist-packages/flask/app.py:1544: in handle_user_exception
    reraise(exc_type, exc_value, tb)
/usr/local/lib/python3.4/dist-packages/flask/_compat.py:33: in reraise
    raise value
/usr/local/lib/python3.4/dist-packages/flask/app.py:1639: in full_dispatch_request
    rv = self.dispatch_request()
/usr/local/lib/python3.4/dist-packages/flask/app.py:1625: in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
source.py:72: in decorated_function
    return f(*args, **kwargs)
source.py:301: in submit
    fh.stream))
store.py:115: in save_file_submission
    crypto_util.encrypt(stf, config.JOURNALIST_KEY, encrypted_file_path)
crypto_util.py:158: in encrypt
    armor=False)
/usr/local/lib/python3.4/dist-packages/gnupg/gnupg.py:972: in encrypt
    stream = _make_binary_stream(data, self._encoding)

thing = <secure_tempfile.SecureTemporaryFile object at 0x7f383018a400>, encoding = 'utf_8', armor = True

    def _make_binary_stream(thing, encoding=None, armor=True):
        """Encode **thing**, then make it stream/file-like.
    
        :param thing: The thing to turn into a encoded stream.
        :rtype: ``io.BytesIO`` or ``io.StringIO``.
        :returns: The encoded **thing**, wrapped in an ``io.BytesIO`` (if
            available), otherwise wrapped in a ``io.StringIO``.
        """
        if _py3k:
            if isinstance(thing, str):
                thing = thing.encode(encoding)
        else:
            if type(thing) is not str:
                thing = thing.encode(encoding)
    
        try:
>           rv = BytesIO(thing)
E           TypeError: 'SecureTemporaryFile' does not support the buffer interface
```

--- a/securedrop/crypto_util.py
+++ b/securedrop/crypto_util.py
@@ -5,6 +5,7 @@ from base64 import b32encode
 
 from Crypto.Random import random
 import gnupg
+from gnupg._util import _is_stream, _make_binary_stream
 import scrypt
 
 import config
@@ -154,6 +155,9 @@ def encrypt(plaintext, fingerprints, output=None):
         fingerprints = [fingerprints, ]
     fingerprints = [fpr.replace(' ', '') for fpr in fingerprints]
 
+    if not _is_stream(plaintext):
+        plaintext = _make_binary_stream(plaintext, "utf_8")
+
     out = gpg.encrypt(plaintext,
                       *fingerprints,
                       output=output,
@@ -165,7 +169,7 @@ def encrypt(plaintext, fingerprints, output=None):
         raise CryptoException(out.stderr)
 
 
-def decrypt(secret, plain_text):
+def decrypt(secret, ciphertext):
     """
     >>> key = genkeypair('randomid', 'randomid')
     >>> decrypt('randomid', 'randomid',
@@ -174,8 +178,7 @@ def decrypt(secret, plain_text):
     'Goodbye, cruel world!'
     """
     hashed_codename = hash_codename(secret, salt=SCRYPT_GPG_PEPPER)
-    return gpg.decrypt(plain_text, passphrase=hashed_codename).data
-
+    return gpg.decrypt(ciphertext, passphrase=hashed_codename).data
 
 if __name__ == "__main__":
     import doctest

--- a/securedrop/tests/test_unit_integration.py
+++ b/securedrop/tests/test_unit_integration.py
@@ -257,8 +257,21 @@ class TestIntegration(unittest.TestCase):
     def test_reply_normal(self):
         self.helper_test_reply("This is a test reply.", True)
 
-    def test_reply_unicode(self):
-        self.helper_test_reply("Teşekkürler", True)
+    def test_unicode_reply_with_ansi_env(self):
+        # This makes python-gnupg handle encoding equivalent to if we were
+        # running SD in an environment where os.getenv("LANG") == "C".
+        # Unfortunately, with the way our test suite is set up simply setting
+        # that env var here will not have the desired effect. Instead we
+        # monkey-patch the GPG object that is called crypto_util to imitate the
+        # _encoding attribute it would have had it been initialized in a "C"
+        # environment. See
+        # https://github.com/freedomofpress/securedrop/issues/1360 for context.
+        old_encoding = crypto_util.gpg._encoding
+        crypto_util.gpg._encoding = "ansi_x3.4_1968"
+        try:
+            self.helper_test_reply("ᚠᛇᚻ᛫ᛒᛦᚦ᛫ᚠᚱᚩᚠᚢᚱ᛫ᚠᛁᚱᚪ᛫ᚷᛖᚻᚹᛦᛚᚳᚢᛗ", True)
+        finally:
+            crypto_util.gpg._encoding = old_encoding
 
     def _can_decrypt_with_key(self, msg, key_fpr, passphrase=None):
         """

